File: mgmt-application/src/main/java/io/quartic/mgmt/MgmtApplication.java
package io.quartic.mgmt;

import io.dropwizard.assets.AssetsBundle;
import io.dropwizard.setup.Bootstrap;
import io.dropwizard.setup.Environment;
import io.quartic.catalogue.api.CatalogueService;
import io.quartic.common.application.ApplicationBase;
import io.quartic.common.healthcheck.PingPongHealthCheck;
import io.quartic.howl.api.HowlClient;
import io.quartic.howl.api.HowlService;

import static io.quartic.common.client.ClientUtilsKt.client;
import static io.quartic.common.client.ClientUtilsKt.userAgentFor;

public class MgmtApplication extends ApplicationBase<MgmtConfiguration> {
    public static void main(String[] args) throws Exception {
        new MgmtApplication().run(args);
    }

    @Override
    public void initializeApplication(Bootstrap<MgmtConfiguration> bootstrap) {
        bootstrap.addBundle(new AssetsBundle("/assets", "/", "index.html"));
    }

    @Override
    public void runApplication(MgmtConfiguration configuration, Environment environment) {
        HowlService howlService = new HowlClient(userAgentFor(MgmtApplication.class), configuration.getHowlUrl());

        environment.healthChecks().register("catalogue", new PingPongHealthCheck(getClass(), configuration.getCatalogueUrl()));

        CatalogueService catalogueService = client(CatalogueService.class, getClass(), configuration.getCatalogueUrl());
        environment.jersey().register(new MgmtResource(catalogueService, howlService, configuration.getDefaultCatalogueNamespace()));
    }
}

File: mgmt-application/src/main/kotlin/io/quartic/mgmt/MgmtApplication.kt
package io.quartic.mgmt

import io.dropwizard.assets.AssetsBundle
import io.dropwizard.setup.Bootstrap
import io.dropwizard.setup.Environment
import io.quartic.bild.api.BildQueryService
import io.quartic.catalogue.api.CatalogueService
import io.quartic.common.application.ApplicationBase
import io.quartic.common.application.TokenAuthConfiguration
import io.quartic.common.auth.TokenGenerator
import io.quartic.common.client.client
import io.quartic.common.client.retrofitClient
import io.quartic.common.client.userAgentFor
import io.quartic.common.healthcheck.PingPongHealthCheck
import io.quartic.howl.api.HowlClient
import io.quartic.mgmt.resource.AuthResource
import io.quartic.mgmt.resource.MgmtResource
import io.quartic.mgmt.resource.UserResource
import io.quartic.registry.api.RegistryService
import io.quartic.registry.api.RegistryServiceClient
import java.time.Duration

class MgmtApplication : ApplicationBase<MgmtConfiguration>() {

    public override fun initializeApplication(bootstrap: Bootstrap<MgmtConfiguration>) {
        bootstrap.addBundle(AssetsBundle("/assets", "/", "index.html"))
    }

    public override fun runApplication(configuration: MgmtConfiguration, environment: Environment) {
        val howl = HowlClient(userAgentFor(javaClass), configuration.howlUrl)
        val catalogue = client<CatalogueService>(javaClass, configuration.catalogueUrl)
        val registry = retrofitClient<RegistryServiceClient>(javaClass, configuration.registryUrl)
        val bild = client<BildQueryService>(javaClass, configuration.bildUrl)

        val tokenGenerator = TokenGenerator(
            configuration.auth as TokenAuthConfiguration,
            configuration.secretsCodec,
            Duration.ofMinutes(configuration.tokenTimeToLiveMinutes.toLong())
        )

        with (environment.jersey()) {
            register(UserResource(configuration.github))
            register(MgmtResource(
                catalogue,
                howl,
                bild,
                registry
            ))
            register(AuthResource(
                configuration.github,
                configuration.cookies,
                configuration.secretsCodec,
                tokenGenerator,
                registry
            ))
        }
        environment.healthChecks().register("catalogue", PingPongHealthCheck(javaClass, configuration.catalogueUrl))
    }

    companion object {
        @JvmStatic fun main(args: Array<String>) = MgmtApplication().run(*args)
    }
}
